### PR TITLE
ADD: MRR2 backend reattempt

### PR DESCRIPTION
**Pull request by rcjackson**
I have restarted the pull request for the MRR2 reader using the current version as a base. I'm hoping that I've made changes that fix the PR. 
- [x] Closes #xxxx
- [x] Tests added
- [x] Changes are documented in `history.md`

### CODE REVIEW

**Comment by rcjackson** [comment]
Ruff is failing on notebooks unrelated to this PR.

**Comment by kmuehlbauer** [comment]
@rcjackson Any idea on the `bus error` at the one nexrad test?

**Comment by rcjackson** [comment]
Memory leak somewhere? I haven’t been able to find it.

From: Kai Mühlbauer ***@***.***>
Date: Tuesday, August 27, 2024 at 1:30 AM
To: openradar/xradar ***@***.***>
Cc: Jackson, Robert ***@***.***>, Mention ***@***.***>
Subject: Re: [openradar/xradar] ADD: MRR2 backend reattempt (PR #198)
@ rcjackson Any idea on the bus error at the one nexrad test? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned. Message ID: <openradar/xradar/pull/198/c2311678517@ github. com>
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd

@rcjackson<https://urldefense.us/v3/__https:/github.com/rcjackson__;!!G_uCfscf7eWS!Y_ThD4cL6cHVVZFproDciAN6Zhh3zHbZjGjkBh00rdXNdZifRZTblHs3Olk6cj_ElibCJCXw1HbCk4qgArYj4Sko$> Any idea on the bus error at the one nexrad test?

File: tests/io/test_metek.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        2.000e00,
+        1.000e01,
+        1.250e02,
+        6.330e02,
+    ]
+)
+
+
+def test_open_average(metek_ave_gz_file):
+    ds = xr.open_dataset(metek_ave_gz_file, engine="metek")
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:153]: Could you try using context manager here?

File: tests/io/test_metek.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    assert "corrected_reflectivity" in ds.variables.keys()
+    assert "velocity" in ds.variables.keys()
+    rainfall = ds["rainfall_rate"].isel(range=0).cumsum() / 60.0
+    np.testing.assert_allclose(rainfall.values[-1], 0.938)
+    np.testing.assert_allclose(ds["reflectivity"].values[0], test_arr_ave)
+    ds.close()
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:159]: ... and remove these two lines?

File: tests/io/test_metek.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def test_open_average_datatree(metek_ave_gz_file):
+    ds = metek.open_metek_datatree(metek_ave_gz_file)
+    assert "corrected_reflectivity" in ds["sweep_0"].variables.keys()
+    assert "velocity" in ds["sweep_0"].variables.keys()
+    rainfall = ds["sweep_0"]["rainfall_rate"].isel(range=0).cumsum() / 60.0
+    np.testing.assert_allclose(rainfall.values[-1], 0.938)
+    np.testing.assert_allclose(ds["sweep_0"]["reflectivity"].values[0], test_arr_ave)
+    ds = None
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:170]: Remove

File: tests/io/test_metek.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def test_open_processed(metek_pro_gz_file):
+    ds = xr.open_dataset(metek_pro_gz_file, engine="metek")
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:174]: Use context manager.

File: tests/io/test_metek.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def test_open_processed_datatree(metek_pro_gz_file):
+    ds = metek.open_metek_datatree(metek_pro_gz_file)
+    assert "corrected_reflectivity" in ds["sweep_0"].variables.keys()
+    assert "velocity" in ds["sweep_0"].variables.keys()
+    rainfall = ds["sweep_0"]["rainfall_rate"].isel(range=0).cumsum() / 360.0
+    np.testing.assert_allclose(rainfall.values[-1], 0.93)
+    np.testing.assert_allclose(ds["sweep_0"]["reflectivity"].values[0], test_arr)
+    ds = None
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:191]: Remove

File: tests/io/test_metek.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def test_open_raw(metek_raw_gz_file):
+    ds = xr.open_dataset(metek_raw_gz_file, engine="metek")
+    assert "raw_spectra_counts" in ds.variables.keys()
+    np.testing.assert_allclose(ds["raw_spectra_counts"].values[0], test_raw)
+    ds.close()
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:198]: Remove two lines

File: tests/io/test_metek.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    ds = None
+
+
+def test_open_raw_datatree(metek_raw_gz_file):
+    ds = metek.open_metek_datatree(metek_raw_gz_file)
+    assert "raw_spectra_counts" in ds["sweep_0"].variables.keys()
+    np.testing.assert_allclose(ds["sweep_0"]["raw_spectra_counts"].values[0], test_raw)
+    ds = None
```
Comment (kmuehlbauer) [inline tests/io/test_metek.py:206]: remove

**Review by kmuehlbauer — CHANGES_REQUESTED** [review]
@rcjackson I've made a couple comments, how we might tackle the issue with the `bus error`.

I've checked the tests on main here:
https://github.com/openradar/xradar/actions/runs/10578167450
and they are working perfectly fine. 

So I'm assuming that the ominous `bus error` is somehow introduced with the changes in this PR, or are revealed by the changes in this PR. 

If you have any other ideas how to tackle this, please comment.

File: tests/conftest.py [hunk 8]
```diff
@@ -96,3 +98,45 @@ def nexradlevel2_bzfile():
...
+def metek_ave_gz_file():
+    fnamei = DATASETS.fetch("0308.ave.gz")
+    fnameo = f"{fnamei[:-3]}"
+    import gzip
+    import shutil
+
+    with gzip.open(fnamei) as fin:
+        with open(fnameo, "wb") as fout:
+            shutil.copyfileobj(fin, fout)
+    os.remove(fnamei)
```
Comment (kmuehlbauer) [inline tests/conftest.py:113]:
```suggestion
```

File: tests/conftest.py [hunk 9]
```diff
@@ -96,3 +98,45 @@ def nexradlevel2_bzfile():
...
+def metek_pro_gz_file():
+    fnamei = DATASETS.fetch("0308.pro.gz")
+    fnameo = f"{fnamei[:-3]}"
+    import gzip
+    import shutil
+
+    with gzip.open(fnamei) as fin:
+        with open(fnameo, "wb") as fout:
+            shutil.copyfileobj(fin, fout)
+    os.remove(fnamei)
```
Comment (kmuehlbauer) [inline tests/conftest.py:127]:
```suggestion
```

File: tests/conftest.py [hunk 10]
```diff
@@ -96,3 +98,45 @@ def nexradlevel2_bzfile():
...
+def metek_raw_gz_file():
+    fnamei = DATASETS.fetch("0308.raw.gz")
+    fnameo = f"{fnamei[:-3]}"
+    import gzip
+    import shutil
+
+    with gzip.open(fnamei) as fin:
+        with open(fnameo, "wb") as fout:
+            shutil.copyfileobj(fin, fout)
+    os.remove(fnamei)
```
Comment (kmuehlbauer) [inline tests/conftest.py:141]:
```suggestion
```

File: tests/io/test_metek.py (hunk 2 above)
Comment (kmuehlbauer) [inline tests/io/test_metek.py:159]:
```suggestion
    with xr.open_dataset(metek_ave_gz_file, engine="metek") as ds:
        assert "corrected_reflectivity" in ds.variables.keys()
        assert "velocity" in ds.variables.keys()
        rainfall = ds["rainfall_rate"].isel(range=0).cumsum() / 60.0
        np.testing.assert_allclose(rainfall.values[-1], 0.938)
        np.testing.assert_allclose(ds["reflectivity"].values[0], test_arr_ave)
```

**Comment by rcjackson** [comment]
@kmuehlbauer  None of the suggestions worked. I even explicitly close the datatree's Dataset to make sure the references clear, but that still doesn't seem to resolve the issue. I'll have to think of more solutions for this later, but for now I'm at a loss.

**Comment by kmuehlbauer** [comment]
@rcjackson This seems like a hard nut to crack, thanks for checking this out. 

Maybe those open references are some hint as to why the error occurred in the first place? I'll have a look at our mrr reader (which is extremely fragile while being based on numpy.genfromtxt).

**Comment by kmuehlbauer** [comment]
@rcjackson We might try to run this sequentially.

Please change here:

https://github.com/openradar/xradar/blob/1b4a535f582de9083debc86860010aa20cb1f98c/.github/workflows/ci.yml#L69

and remove `-n auto` from the call.

Another thing to look after is this error:

```python
  Traceback (most recent call last):
    File "/home/runner/work/xradar/xradar/xradar/io/backends/metek.py", line 447, in close
      for k in self._data.keys():
               ^^^^^^^^^^
  AttributeError: 'MRR2File' object has no attribute '_data'. Did you mean: 'data'?
```

which was issued here: https://github.com/openradar/xradar/actions/runs/10600167596/job/29376980199?pr=198

It seems, that one CI run finishes and all other raise with this error. Strange!

**Comment by rcjackson** [comment]
> @rcjackson We might try to run this sequentially.
> 
> Please change here:
> 
> https://github.com/openradar/xradar/blob/1b4a535f582de9083debc86860010aa20cb1f98c/.github/workflows/ci.yml#L69
> 
> and remove `-n auto` from the call.
> 
> Another thing to look after is this error:
> 
> ```python
>   Traceback (most recent call last):
>     File "/home/runner/work/xradar/xradar/xradar/io/backends/metek.py", line 447, in close
>       for k in self._data.keys():
>                ^^^^^^^^^^
>   AttributeError: 'MRR2File' object has no attribute '_data'. Did you mean: 'data'?
> ```
> 
> which was issued here: https://github.com/openradar/xradar/actions/runs/10600167596/job/29376980199?pr=198
> 
> It seems, that one CI run finishes and all other raise with this error. Strange!

Running it sequentially seems to get it working, though slows down the testing a bit. Azure stores are typically memory limited at 1 GB, so I bet it was running in parallel and going up against the memory limit.

**Comment by kmuehlbauer** [comment]
Ok, so this looks like a memory leak somewhere.

This will need some profiling, I guess.

**Review by kmuehlbauer — APPROVED** [review]

**Pull request merged**